modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairEntryProcessor.java

```diff
@@ -0,0 +1,204 @@
+         * Fail, not enough information for modification.
+         */
+        FAIL,
+
+        /**
+         * Value was modified from other thread. Result same the success.
+         */
+        CONCURRENT_MODIFICATION
+    }
+
+    /**
+     * @param val Value.
+     * @param data Data.
+     * @param rmvQueueMaxSize Remove queue max size.
+     * @param forceRepair Force repair.
+     * @param startTopVer Start topology version.
+     */
+    @SuppressWarnings("AssignmentOrReturnOfFieldWithMutableType")
+    public RepairEntryProcessor(
+        Object val,
+        Map<UUID, VersionedValue> data,
+        long rmvQueueMaxSize,
+        boolean forceRepair,
+        AffinityTopologyVersion startTopVer) {
+        this.val = val;
+        this.data = data;
+        this.rmvQueueMaxSize = rmvQueueMaxSize;
+        this.forceRepair = forceRepair;
+        this.startTopVer = startTopVer;
+    }
+
+    /**
+     * Do repair logic.
+     *
+     * @param entry Entry to fix.
+     * @param arguments Arguments.
+     * @return {@link RepairStatus} looks at description of this class.
+     * @throws EntryProcessorException If failed.
+     */
+    @SuppressWarnings("unchecked")
+    @Override public Object process(MutableEntry entry, Object... arguments) throws EntryProcessorException {
+        GridCacheContext cctx = cacheContext(entry);
+        GridCacheVersion currKeyGridCacheVer = keyVersion(entry);
+
+        if (topologyChanged(cctx, startTopVer))
+            throw new EntryProcessorException("Topology version was changed");
+
+        UUID locNodeId = cctx.localNodeId();
+        VersionedValue versionedVal = data.get(locNodeId);
+
+        if (versionedVal != null) {
+            if (currKeyGridCacheVer.compareTo(versionedVal.version()) == 0) {
+                if (val == null)
+                    entry.remove();
+                else
+                    entry.setValue(val);
+
+                return RepairStatus.SUCCESS;
+            }
+            else
+                return RepairStatus.CONCURRENT_MODIFICATION;
+        }
+        else {
+            if (currKeyGridCacheVer.compareTo(new GridCacheVersion(0, 0, 0)) == 0) {
+                long recheckStartTime = minValue(VersionedValue::recheckStartTime);
+
+                boolean inEntryTTLBounds =
+                    (System.currentTimeMillis() - recheckStartTime) < Long.getLong(IGNITE_CACHE_REMOVED_ENTRIES_TTL, 10_000);
+
+                // Min available update counter for the key at all nodes.
+                // It just fast solution for null value problem. We should use other way to fix it (versionedVal.updateCounter()).
+                long minUpdateCntr = minValue(VersionedValue::updateCounter);
+                long currUpdateCntr = updateCounter(cctx, entry.getKey());
+
+                boolean inDeferredDelQueueBounds = ((currUpdateCntr - minUpdateCntr) < rmvQueueMaxSize);
+
+                // Remove it after fixes: https://ggsystems.atlassian.net/browse/GG-27419
```

Comment:
This comment will never be found. GG-27419 doesn't refers to this place and this comment doesn't looks like TODO-comment.